### DISCUSSION TITLE
RFC: Standardize v-model Property Names

### DISCUSSION THREAD

**Discussion by seggewiss** [shopware/platform#5654] [RFC / Proposal]
# RFC: Standardize v-model Property Names

## Summary
Propose standardization of v-model property names across the codebase to align with Vue.js conventions, using `modelValue` as the standard property name and `update:model-value` as the event name.

## Background
Currently, the codebase uses various property names for v-model bindings:
- `value`
- `modelValue`
- `entityCollection`
- (potentially others)

This inconsistency makes the codebase harder to maintain and violates Vue.js best practices.

## Motivation
- Align with Vue.js standards and best practices
- Improve code consistency
- Reduce cognitive overhead for developers
- Make component interfaces more predictable
- Easier onboarding for new team members

### Standard Convention
All components using v-model should:
1. Use `modelValue` as the prop name
2. Use `update:model-value` as the event name

### Example Implementation
```vue
<!-- Before -->
<template>
  <input 
    :value="value"
    @input="$emit('input', $event.target.value)"
  />
</template>

<script>
export default {
  props: ['value']
}
</script>

<!-- After -->
<template>
  <input 
    :value="modelValue"
    @input="$emit('update:model-value', $event.target.value)"
  />
</template>

<script>
export default {
  props: {
    modelValue: {
      type: [String, Number],
      required: true
    }
  }
}
</script>
```

### Phase 1: Audit -> Shopware version 6.6
- Create inventory of all components using v-model
- Deprecate all v-model properties that don't use `modelValue` already for Shopware version 6.8
- Deprecate all `emits` of related v-model events for Shopware version 6.8

### Phase 2: Implementation -> Shopware Version 6.7
1. Create new prop with `modelValue`
2. Emit both old and new events during transition
3. Update component documentation
4. Update consuming components
5. Add ESLint rule to enforce naming convention

### Phase 3: Clean up -> Shopware Version 6.8
1. Remove deprecated v-model properties
2. Remove deprecated events

## Drawbacks
- Breaking change requiring updates to consuming components
- Migration effort required
- Temporary code complexity during transition period

## Alternatives Considered
1. Keep current mixed naming convention
   - Rejected due to maintenance overhead and deviation from standards
3. Use different standard property name
   - Rejected as `modelValue` is Vue.js convention

## Additional Context
This change aligns with Vue 3's composition API conventions and will make future Vue version migrations easier.

## Timeline
- Shopware 6.6: Add deprecations for old v-model's and events
- Shopware 6.7: Add new convention
- Shopware 6.8: Remove old v-model's and events

## References
- [Vue.js v-model documentation](https://vuejs.org/guide/components/v-model.html)
- [Vue Style Guide](https://vuejs.org/style-guide/)

**Comment by Haberkamp** [comment]
This is great. What do you think about creating an ESLint rule for that? This way we have an overview about how many v-models are violating the guideline. Other projects can also use that rule, that way we have that consistency in all other projects. 

Currently we're planning to create a set of ESLint and Stylelint rules for Meteor, which we will use for Meteor itself. Though it's also intented that other projects use that. The rule would be a good fit for this new project.

**Reply by seggewiss** [reply]
I like the idea having an Eslint rule for that, but I think this will be difficult as every prop with Vue3 can be bind to via v-model and the model definition with the prop and the event no longer exists. 🤔

**Comment by niklaswolf** [comment]
I like it, using the standards/conventions is most often a good thing :)

**Comment by cyl3x** [comment]
I like standardising things. Is there a reason for using `modelValue` instead of `value`, which vue has been using for a long time?

**Reply by niklaswolf** [reply]
That's the default/convention with defineModel, see https://vuejs.org/guide/components/v-model.html#under-the-hood 😊

**Comment by seggewiss** [comment]
This is still something we want to do in the long run, but this will be achieved using different strategies such as decoupling the admin more via the SDK and using more and more meteor components.
